### ISSUE TITLE
Discussion: Best practice redux + redux-saga application project structure

### ISSUE THREAD

**Issue by aight8** [redux-saga/redux-saga#799]
I write this post because I'm not happy with my current project file structure and I want to give a board to discuss this issue.
There are tons of recommendations out there, the most of them are only **redux** specific. In fact, when using redux with redux-saga it should be integrated as one whole.

Let's discuss about the best practice or package suggestions. Otherwise I will create a package which solves that. Furthermore it's not only about a package which helps, it's a discussion about best practice about **redux + redux-saga architecture**.

Let's start:
---

Normally in a traditional store in a basic application you implement following steps when you add a sub storage to redux (using redux, redux-actions and redux-saga).

- createMyStore.js (1. append )
- allReducers.js (1. import reducerDomainA.js and append the reducer to combineReducers)
    - reducerDomainA.js (1. return handlers wrapped in handleActions)
    - actionCreatorDomainA.js (1. import actionDomainA.js and use action constants 2. implement and export some action creators)
    - actionsDomainA.js (1. implement and export some actions)
    - sagasDomainA.js (1. implement and export some sagas)
- someSmartComponent.js (1. import actionCreatorDomainA.js and bind it with bindActionCreators)

**Don't forget the custom selector functions (optionally with reselect)!**

I’m not happy with this widely suggested solution because it clutters the application literally in **million** parts. I using this now for several month, but I don't get to like it - however I like the architecture itself.

Some focus points are:
- It’s really difficult to see an overview about the flow because of the file structure
- Actions are normally handled by a **saga** OR a **reducer**. You don’t know which one handles it - you must look after the reducers and saga's.
- I just don’t like this in my smart component „import * as actionsDomainA from 'actions/actionsDomainA.js'“

What I would like to see:
- Domain based file structure (bunch domain specific parts in a class but make them accessible)
- Less definition of all actions
- Use js classes to bundle domain specific methods and specify a rough structure
- Testable
- **NO MAGIC** - We are programers and no wizards.

- Domain specific actions

### Current packages which try to increase the productivity:
- https://github.com/pauldijou/redux-act
   - Uses magic, just solves a small part of this list, action types are generated

---

So this is just a quick draft for brainstorming how a domain specific class could be:

```JS
export default class Auth {
  (static) defaultState = {};

  (static) actionCreators = {
    setAuth: null,
    login(email: string, password:string) {
      return { type: 'LOGIN_USER', email, password };
    }
  };

  SET_AUTH (state, action) {
    return action.payload;
  }

  * LOGIN_USER (action, options) {

  }
}
```

#### Current thoughts (about the draft)
- The domain specific class should provide **reducers**, **sagas**, **action creators**
- Normal class methods are seen as reducers which handle actions of the same name than the method
- Generator class methods are seen as sagas which handle actions of the same name than the method

#### Ideas:
- The above class could extend another base class which gives basic functionalities.

#### Open points:
- Current draft uses unconventional method names
- All domain classes are registered once by a central module which load's it to the application core and make it also accessible
- Member properties static or not (for defaultState etc.)?
- Core should be still extendable with domain **unspecific** reducers, sagas etc.

Let's discuss now about both: Best practices and the ideal integration of redux and redux-saga - which could end up in a new package.

**Comment by granmoe** [comment]
Have you heard of the ducks pattern? What you're describing sounds very similar to it: everything in one place, but divided up by domain: https://github.com/erikras/ducks-modular-redux.

I just use this and put my saga code for each duck into the duck itself. Just a slight modification on the ducks pattern. Then within each duck, I do `export sagas = [saga1, saga2]`, and wherever I create my rootReducer I import them via `import { sagas as paymentSagas } from 'ducks/payment'`. Then I just map these in the root reducer.

**Comment by granmoe** [comment]
Also, some of this is dependent on your use case, and some of it is bikeshedding. I don't think it's necessarily a good idea to prescribe one "silver bullet" structure for all redux/redux-saga apps :-/

However, there are many, many good options. You may want to check out the many relevant redux articles and FAQs from Mark Erikson: https://github.com/markerikson/react-redux-links

**Comment by granmoe** [comment]
Any advantage to using classes instead of just plain functions (like in the ducks pattern)?

**Comment by aight8** [comment]
I don't see this repository with the proposal, but thank for the input.

The classes are sugar to bundle the different parts together. Otherwise you just export several functions, but how the core knows which export is what. Additionally a base class could provide a good basic functionality. I really ask here: There are classes now, when it make it things cleaner why not use them. Wasn't that the basic thought of classes?

I am already created a common saga generator which handle unhandled exception and uses takeLatest. Additionally provide callback methods etc.
But this really just solves a small part.

I definitely don't want to prescribe a silver bullet. But I want to simplify very common patterns, and improve readability.

**Comment by Andarist** [comment]
Well, Im not really sure about what 'saga generator' are you talking about and what problem really it would solve. On the one hand you say we are not wizards and on the other hand I feel like you propose making some magical generator.

>  Even React uses classes for components with different predefined lifecycle methods and (static) property names which can be implemented.

For now... from what I know there is a plan to move away from classes in React, aint sure if it succeeds, but it might. It could possibly even solve some issues (for example with hot reloading), and it is not an attempt for ditching classes just for the sake of... ditching them ;)

From my point of view classes should be used really sparingly, they sneaked to JS by accident to make some comfort for more classy-based languages like Java and there is literally no single thing they actually solves for JS developers, everything they can do could be already done, so its just a sugar, but quite a dangerous one.

I think class-based object oriented programming tempts people to couple things with each other and creates an inheritance hell. If you can avoid this, by simply not using them (unless you really know what you are doing, and you wont use inheritance), its a good way for keeping yourself constrained in a way. Which can lead you more to composition, than to inheritance, which in the end will make you a better programmer.

Ofc thats just my opinion (and many other ;) ), so dont take it for granted if you do not like it. Just my 2 cents.

**Comment by kairome** [comment]
Regarding this 
> action types are generated and could be not reused in specific uses cases when the action name must be known outside of the app

a little update, you can now get the type by using `getType()` method on the action itself.
https://github.com/pauldijou/redux-act#action-creator

**Comment by Andarist** [comment]
I believe it was possible long time ago - just with different function: `someAction.toString()`

**Comment by wisn** [comment]
Have you take a look on [Ignite RN boilerplate](https://github.com/infinitered/ignite-example-login/tree/master/App/Sagas)?

**Comment by milieu** [comment]
re: @granmoe class components vs function components

> You may not use the ref attribute on function components because they don’t have instances
-- [Refs and the DOM](https://reactjs.org/docs/refs-and-the-dom.html#refs-and-function-components)

this turned out relevant for me while using react-window for performance improvements, and needing to allow the library to forward refs for scroll control

**Comment by saboya** [comment]
@milieu This is not true anymore since `react-16.3` with `createRef`/`forwardRef` and now, `react-16.8` with the `React.useRef` hook.
